### PR TITLE
Wire up subnav

### LINKED ISSUE #505: Build Feeder Subnav
After #503.  Build out the [sub nav bar](https://www.figma.com/proto/Yewv83Q8btHcaOBWCtGgKs/PRX-Dovetail-with-Navigation?page-id=80%3A2947&node-id=2934%3A80025&viewport=798%2C-3032%2C0.25&scaling=min-zoom&starting-point-node-id=80%3A3420&hide-ui=1), which has some dynamic content to whatever page you're on.

image

- [x] Make it _fairly_ bootstrap-ish, and responsive.
- [x] Add a before action to application controller (or whatever we're using as the parent to the HTML controllers).
  - Load the complete list of policy-readable podcasts
  - Order by title here
  - Assign to a less-conflict-prone variable like `@subnav_podcasts`
  - Check the current route to see if one of them should be selected (TODO: what about `/episodes/1234` type paths?)
  - Links should go to the `/podcasts/1234` "show" path
- [x] Make the search box work, but frontend only.  Stimulus!
- [x] Check if there is any content for the subnav - render them only if they exist.

### PR DESCRIPTION
**Pull request by cavis**
Closes #505.

Might be slightly overkill, but I opted to make the dropdown a [lazy loaded frame](https://turbo.hotwired.dev/reference/frames#lazy-loaded-frame).  So it only queries the db after you open it.

After typing in the search box, it makes a POST for a turbo-stream, which replaces the list-of-5-podcasts.

### CODE REVIEW

File: app/helpers/podcasts_helper.rb [hunk 1]
```diff
@@ -29,4 +29,11 @@ def itunes_summary(model)
       sanitize_links_only(model.description || "")
     end
   end
+
+  def podcast_settings_active?
+    parts = request.path.split("/").select(&:present?)
+
+    # don't include podcast#show, or any episodes/feeds paths
+    parts[0] == "podcasts" && parts[2] && parts[2] != "episodes" && parts[2] != "feeds"
```
Comment (cavis) [inline app/helpers/podcasts_helper.rb:37]: The hierarchy of podcast routes doesn't _quite_ line up with when the `Settings` subnav link is active.  Have to filter for:

- `podcasts/123` the "podcast dashboard" thing
- `podcasts/123/episodes` listing episodes
- `podcasts/123/feeds` listing feeds

File: app/javascript/controllers/search_controller.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    clearTimeout(this.timeout)
+    this.timeout = setTimeout(() => {
+      this.submitForm()
+    }, 200)
+  }
+
+  // NOTE: form.submit() doesn't seem to play well with Turbo at the moment -
+  // so just click the hidden submit button
+  submitForm() {
+    this.submitTarget.click()
```
Comment (cavis) [inline app/javascript/controllers/search_controller.js:20]: I double checked ... and indeed calling `this.element.submit()` on the `<form>` still doesn't work with Turbo.  It submits a non-turbo-stream-`Accept` request.  Which isn't what I want.

File: app/policies/application_policy.rb [hunk 3]
```diff
@@ -32,4 +32,17 @@ def account_id
   def account_id_was
     nil
   end
+
+  class Scope
+    attr_reader :token, :scope
```
Comment (cavis) [inline app/policies/application_policy.rb:37]: Ported from Augury.  But... the pattern in Feeder seems to be calling it a `token` vs a `user`.  So stuck with that here.  (It's the same `Prx::Auth::Token` so doesn't really matter).

File: app/policies/podcast_policy.rb [hunk 4]
```diff
@@ -1,4 +1,8 @@
 class PodcastPolicy < ApplicationPolicy
+  def new?
+    token&.authorized_account_ids(:podcast_create).present?
```
Comment (cavis) [inline app/policies/podcast_policy.rb:3]: The `create?` method requires the Podcast have an account set on it already.  So needed a check for "can you create a podcast for _any_ account ids?".

File: app/policies/podcast_policy.rb [hunk 5]
```diff
@@ -20,4 +24,12 @@ def account_id
   def account_id_was
     resource.account_id_was
   end
+
+  class Scope < Scope
+    def resolve
+      # TODO: this is hacky
+      uris = token.authorized_account_ids(:read_private).map { |id| "/api/v1/accounts/#{id}" }
```
Comment (cavis) [inline app/policies/podcast_policy.rb:31]: I feel like someday (after deprecating CMS)... we should go through and change all these API-account-strings to just integers.

File: app/views/podcast_switcher/_dropdown.html.erb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+          <%= f.label :q, t(".search") %>
+        </div>
+        <%= f.submit class: "d-none", data: {search_target: "submit"} %>
+      <% end %>
+    </li>
+
+    <%# lazy-loaded list of podcasts %>
+    <%= turbo_frame_tag "podcast-switcher-list", src: main_app.podcast_switcher_path, loading: "lazy" do %>
+      <li class="d-flex justify-content-center">
+        <div class="spinner-border text-primary m-5" role="status"></div>
```
Comment (cavis) [inline app/views/podcast_switcher/_dropdown.html.erb:34]: Shows a spinner while the list-of-5 lazy loads.  Saves us querying the DB for a hidden dropdown every page load.

**Review by brandonhundt — COMMENTED** [review]
@cavis Made one minor styling change for handling when an podcast image isn't present. 

The rest of the functionality is great! Thanks!!! 

Ready for Code Review.

File: app/views/podcast_switcher/_dropdown.html.erb [hunk 7]
```diff
@@ -1,19 +1,18 @@
-<div class="prx-podcast-switcher dropdown" data-controller="focus">
+<div class="prx-podcast-switcher | dropdown" data-controller="focus">
```
Comment (brandonhundt) [inline app/views/podcast_switcher/_dropdown.html.erb:1]: @rpeterman-gp What's the rationale behind the `|`, just curious?

File: app/views/podcast_switcher/_dropdown.html.erb (hunk 6 above)
Comment (svevang) [reply app/views/podcast_switcher/_dropdown.html.erb:33]: Screen Shot 2023-02-02 at 10 19 30 AM

Seeing some titles hanging off the edge here.

**Review by svevang — APPROVED** [review]

File: app/policies/application_policy.rb (hunk 3 above)
Comment (cavis) [reply app/policies/application_policy.rb:37]: It seems they already are?  I've actually been using `current_user` in other places (outside of the Policies).

**Pull request merged**